server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java

```diff
@@ -0,0 +1,204 @@
+                                    if (termCount == 0) {
+                                        // Brand new term, save into map
+                                        map.put(val, 1L);
+                                        long bucketOrdinal = bucketOrds.add(val);
+                                        if (bucketOrdinal < 0) { // already seen
+                                            bucketOrdinal = - 1 - bucketOrdinal;
+                                            collectExistingBucket(subCollectors, docId, bucketOrdinal);
+                                        } else {
+                                            collectBucket(subCollectors, docId, bucketOrdinal);
+                                        }
+                                    } else {
+                                        // We've seen this term before, but less than the threshold
+                                        // so just increment its counter
+                                        if (termCount < maxDocCount) {
+                                            // TODO if we only need maxDocCount==1, we could specialize
+                                            // and use a bitset instead of a counter scheme
+                                            map.put(val, termCount + 1);
+                                        } else {
+                                            // Otherwise we've breached the threshold, remove from
+                                            // the map and add to the bloom filter
+                                            map.remove(val);
+                                            bloom.put(val);
+                                            numDeleted += 1;
+
+                                            if (numDeleted > GC_THRESHOLD) {
+                                                gcDeletedEntries();
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                            previous = val;
+                        }
+                    }
+                }
+            }
+        };
+    }
+
+    protected void gcDeletedEntries() {
+        boolean hasDeletedEntry = false;
+        LongHash newBucketOrds = new LongHash(1, context.bigArrays());
+        try (LongHash oldBucketOrds = bucketOrds) {
+
+            long[] mergeMap = new long[(int) oldBucketOrds.size()];
+
+            for (int i = 0; i < oldBucketOrds.size(); i++) {
+                long oldKey = oldBucketOrds.get(i);
+                long newBucketOrd = -1;
+
+                // if the key still exists in our map, reinsert into the new ords
+                if (map.containsKey(oldKey)) {
+                    newBucketOrd = newBucketOrds.add(oldKey);
+                } else {
+                    // Make a note when one of the ords has been deleted
+                    hasDeletedEntry = true;
```

Comment:
To make sure this GC is working correctly I wonder if it's worth having a counter here and then checking the counter value is the same as the numDeleted that we expect at the end of this for loop? Another option would be to initialise the variable to `numDeleted` and decrement it here ensuring it reaches 0.